index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,198 @@
+	lists, err := fieldValues.fetchLists(fieldValueFilter, isRegexp)
+	if err != nil {
+		return nil, err
+	}
+
+	// union all the postingsList(s)
+	unionedList := t.opts.PostingsListPool().Get()
+	for _, ids := range lists {
+		unionedList.Union(ids)
+	}
+
+	return unionedList, nil
+}
+
+func (t *simpleTermsDictionary) getOrAddFieldName(fieldName string) *fieldValuesMap {
+	// cheap read lock to see if it already exists
+	t.fieldNamesLock.RLock()
+	fieldValues, ok := t.fieldNames[fieldName]
+	t.fieldNamesLock.RUnlock()
+	if ok {
+		return fieldValues
+	}
+
+	// acquire write lock and create
+	t.fieldNamesLock.Lock()
+	fieldValues, ok = t.fieldNames[fieldName]
+
+	// check if it's been created since we last acquired the lock
+	if ok {
+		t.fieldNamesLock.Unlock()
+		return fieldValues
+	}
+
+	fieldValues = newFieldValuesMap(t.opts)
+	t.fieldNames[fieldName] = fieldValues
+	t.fieldNamesLock.Unlock()
+	return fieldValues
+}
+
+type fieldValuesMap struct {
+	sync.RWMutex
+
+	opts Options
+
+	// fieldValue -> postingsList
+	values map[string]segment.PostingsList
+}
+
+func newFieldValuesMap(opts Options) *fieldValuesMap {
+	return &fieldValuesMap{
+		opts:   opts,
+		values: make(map[string]segment.PostingsList),
+	}
+}
+
+func (f *fieldValuesMap) addDocIDForValue(value string, i segment.DocID) error {
+	// try read lock to see if we already have a postingsList for the given value.
+	f.RLock()
+	pid, ok := f.values[value]
+	f.RUnlock()
+
+	// we have a postingsList, mark the docID and move on.
+	if ok {
+		pid.Insert(i)
+		return nil
+	}
+
+	// postingsList doesn't exist, time to acquire write lock
+	f.Lock()
+	pid, ok = f.values[value]
+
+	// check if it's been created since we released lock
+	if ok {
+		f.Unlock()
+		pid.Insert(i)
+		return nil
+	}
+
+	// create new posting id for the term, and insert into fieldValues
+	pid = f.opts.PostingsListPool().Get()
+	f.values[value] = pid
+	f.Unlock()
+	pid.Insert(i)
+	return nil
+}
+
+func (f *fieldValuesMap) fetchLists(valueFilter []byte, regexp bool) ([]segment.PostingsList, error) {
+	// special case when we're looking for an exact match
+	if !regexp {
+		return f.fetchExact(valueFilter), nil
+	}
+
+	// otherwise, we have to iterate over all known values
+	pred, err := newRegexPredicate(valueFilter)
+	if err != nil {
+		return nil, err
+	}
+	var postingsLists []segment.PostingsList
+	f.RLock()
+	for value, list := range f.values {
+		if pred(value) {
+			postingsLists = append(postingsLists, list)
+		}
+	}
+	f.RUnlock()
+
+	return postingsLists, nil
+}
+
+func (f *fieldValuesMap) fetchExact(valueFilter []byte) []segment.PostingsList {
+	f.RLock()
+	pid, ok := f.values[string(valueFilter)]
+	f.RUnlock()
+	if !ok {
+		return nil
+	}
+	return []segment.PostingsList{pid}
+}
+
+func newRegexPredicate(valueFilter []byte) (valuePredicate, error) {
+	filter := string(valueFilter)
+	re, err := regexp.Compile(filter)
+	if err != nil {
+		return nil, err
+	}
+
+	return re.MatchString, nil
```

Comment:
So we're not using trigrams for v1?

Reply:
We will, #7 is already in the pipeline :) Holding off adding that here to keep PRs small.

Reply:
Makes sense, ok cool.